edx_sga/sga.py

```diff
@@ -0,0 +1,530 @@
+        state['annotated_timestamp'] = _now().strftime(
+            DateTime.DATETIME_FORMAT
+        )
+        path = self._file_storage_path(sha1, filename)
+        if not default_storage.exists(path):
+            default_storage.save(path, File(upload.file))
+        module.state = json.dumps(state)
+        module.save()
+        return Response(json_body=self.staff_grading_data())
+
+    @XBlock.handler
+    def download_assignment(self, request, suffix=''):
+        answer = self.get_submission()['answer']
+        path = self._file_storage_path(answer['sha1'], answer['filename'])
+        return self.download(path, answer['mimetype'], answer['filename'])
+
+    @XBlock.handler
+    def download_annotated(self, request, suffix=''):
+        path = self._file_storage_path(
+            self.annotated_sha1,
+            self.annotated_filename,
+        )
+        return self.download(
+            path,
+            self.annotated_mimetype,
+            self.annotated_filename
+        )
+
+    @XBlock.handler
+    def staff_download(self, request, suffix=''):
+        require(self.is_course_staff())
+        submission = self.get_submission(request.params['student_id'])
+        answer = submission['answer']
+        path = self._file_storage_path(answer['sha1'], answer['filename'])
+        return self.download(path, answer['mimetype'], answer['filename'])
+
+    @XBlock.handler
+    def staff_download_annotated(self, request, suffix=''):
+        require(self.is_course_staff())
+        module = StudentModule.objects.get(pk=request.params['module_id'])
+        state = json.loads(module.state)
+        path = self._file_storage_path(
+            state['annotated_sha1'],
+            state['annotated_filename']
+        )
+        return self.download(
+            path,
+            state['annotated_mimetype'],
+            state['annotated_filename']
+        )
+
+    def download(self, path, mimetype, filename):
+        BLOCK_SIZE = (1 << 10) * 8  # 8kb
+        file = default_storage.open(path)
+        app_iter = iter(partial(file.read, BLOCK_SIZE), '')
+        return Response(
+            app_iter=app_iter,
+            content_type=mimetype,
+            content_disposition="attachment; filename=" + filename)
+
+    @XBlock.handler
+    def get_staff_grading_data(self, request, suffix=''):
+        require(self.is_course_staff())
+        return Response(json_body=self.staff_grading_data())
+
+    @XBlock.handler
+    def enter_grade(self, request, suffix=''):
+        require(self.is_course_staff())
+        module = StudentModule.objects.get(pk=request.params['module_id'])
+        state = json.loads(module.state)
+        score = int(request.params['grade'])
+        if self.is_instructor():
+            uuid = request.params['submission_id']
+            submissions_api.set_score(uuid, score, self.max_score())
+        else:
+            state['staff_score'] = score
```

Comment:
In this particular case, if the staff person creating the score is not the instructor, the score is not meant to be final, but is subject to review by the instructor.  This mechanism is storing the staff assigned score for later review, not setting it on the submission.